### PR TITLE
Catalog Product: fix admin product options saving on store level

### LINKED ISSUE #29572: Product custom option delete while edit product programmatically
### Preconditions (*)

1. Magento 2.3.0
2. php 7.2.3

### Steps to reproduce (*)

1. Add custom options to any product. 
2. Update product SKU programmatically (custom script or custom module) that have custom options, for instance using following code:
```php
$productCollectionFactory = $objectManager->get('\Magento\Catalog\Model\ResourceModel\Product\CollectionFactory');
$storeId = 10;
$productCollection = $productCollectionFactory->create()->addAttributeToSelect('*')->addStoreFilter($storeId);
foreach ($productCollection as $product) {
    $sku = $product->getSku();
    $product->setSku('UpdatedSku-'.$sku);
    $product->save();
}
```
3.  check the same product in the backend

### Expected result (*)

1. The custom options should be unchanged/undeleted

### Actual result (*)

1. Custom option deletes / no more available
---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without the workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look, and feel, “quality” or “usability”._

### LINKED ISSUE #29581: [Issue] Catalog Product: fix admin product options saving on store level
This issue is automatically created based on existing pull request: magento/magento2#29349: Catalog Product: fix admin product options saving on store level

---------

**Preconditions:**

Configure Magento in multi-store mode.
Create a website with 2 store views in it.

**Steps to reproduce:**

1. Create a product in Admin panel that supports custom options.
1.1. Assign a product to the website.
1.2. Add/import some custom options.
2. Save the product.
3. Change the config scope to store view level.
4. Change some of the attributes available on store view (e.g. price).
5. Save the product (don't open Custom Options tab).
6. Open Custom Options tab.

**Actual Result:** :heavy_multiplication_x: Custom options are re-defined on store view level.

Screenshot from 2020-08-31 13-20-35

**Expected Result:** :heavy_check_mark: Custom options are not re-defined on store view level since nothing was actually changed.

Screenshot from 2020-08-31 13-26-14

### Questions or comments

BTW is it OK keeping a store view custom options and options values data in DB after a product is being unassigned from the website related to the store view?

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by oleksii-lisovyi**
### Description (*)

The PR fixed the issue with custom options rewriting on store level without actual changes in case of using Magento in multi-store mode.

### Related Pull Requests
https://github.com/magento/partners-magento2b2b/pull/418

### Fixed Issues (if relevant)

Fixes #29581
Fixes #29572 

### Manual testing scenarios (*)

**Prerequisites**
1. Configure Magento in multi-store mode.
2. Create a website with 2 store views in it.

**Steps to reproduce**
1. Create a product in Admin panel that supports custom options. 
1.1. Assign a product to the website.
1.2. Add/import some custom options.
2. Save the product.
3. Change the config scope to store view level.
4. Change some of the attributes available on store view (e.g. price).
5. Save the product (don't open Custom Options tab).
6. Open Custom Options tab.

**Expected result:**
Custom options are not re-defined on store view level since nothing was actually changed.

**Actual result:**
Custom options are re-defined on store view level.

### Questions or comments

BTW is it OK keeping a store view custom options and options values data in DB after a product is being unassigned from the website related to the store view?

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#29581: Catalog Product: fix admin product options saving on store level

### CODE REVIEW

**Comment by oleksii-lisovyi** [comment]
@magento run all tests

**Comment by oleksii-lisovyi** [comment]
@magento run all tests

**Comment by oleksii-lisovyi** [comment]
@magento run all tests

File: app/code/Magento/Catalog/Test/Unit/Model/Product/Option/SaveHandlerTest.php [hunk 1]
```diff
@@ -40,8 +40,9 @@ protected function setUp(): void
     {
         $this->entity = $this->getMockBuilder(Product::class)
             ->disableOriginalConstructor()
+            ->setMethods(['getOptionsSaved', 'getCanSaveCustomOptions', 'getOptions', 'dataHasChangedFor', 'getSku'])
             ->getMock();
-        $this->optionMock = $this->getMockBuilder(Option::class)
+        $this->option = $this->getMockBuilder(Option::class)
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Test/Unit/Model/Product/Option/SaveHandlerTest.php]: best practice is to have "mock" suffix for cases when object is really mock. Please revert renaming

File: app/code/Magento/Catalog/Test/Unit/Model/Product/Option/SaveHandlerTest.php [hunk 2]
```diff
@@ -53,8 +54,10 @@ protected function setUp(): void
 
     public function testExecute()
     {
-        $this->optionMock->expects($this->any())->method('getOptionId')->willReturn(5);
-        $this->entity->expects($this->once())->method('getOptions')->willReturn([$this->optionMock]);
+        $this->entity->expects($this->once())->method('getOptionsSaved')->willReturn(false);
+        $this->entity->expects($this->once())->method('getCanSaveCustomOptions')->willReturn(true);
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Test/Unit/Model/Product/Option/SaveHandlerTest.php]: Is there any reasons to check that method was ran exactly one time? Are those methods slow?

File: app/code/Magento/Catalog/view/adminhtml/web/js/product/form/fieldset/custom-options.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    return Component.extend({
+        /**
+         * Handler of the "opened" property changes.
+         *
+         * @param {Boolean} isOpened
+         */
+        onVisibilityChange: function (isOpened) {
+            this._super(isOpened);
+
+            // Once the Custom Options tab is opened change "affect_product_custom_options" value to true
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/view/adminhtml/web/js/product/form/fieldset/custom-options.js]:
```suggestion
            // Once the Custom Options tab is opened change "affect_product_custom_options" value to 1
```

File: app/code/Magento/Catalog/view/adminhtml/web/js/product/form/fieldset/custom-options.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+            // in order custom options be processed on backend.
+            if (this.opened()) {
+                uiRegistry.get('product_form.product_form.custom_options.affect_product_custom_options').value(1);
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/view/adminhtml/web/js/product/form/fieldset/custom-options.js]: Would be good to use async way for retrieving this component, like described in https://github.com/magento/magento2/blob/9544fb243d5848a497d4ea7b88e08609376ac39e/app/code/Magento/Ui/view/base/web/js/lib/registry/registry.js#L342-L346

```suggestion
                uiRegistry.get('product_form.product_form.custom_options.affect_product_custom_options', function(component) {
                    component.value(1);
                });
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by ihor-sviziev** [comment]
@oleksii-lisovyi also there is integration test failures, please check them

File: app/code/Magento/Catalog/Test/Unit/Model/Product/Option/SaveHandlerTest.php (hunk 2 above)
Comment (oleksii-lisovyi) [reply app/code/Magento/Catalog/Test/Unit/Model/Product/Option/SaveHandlerTest.php]: So what is the best way to check whether the method was called on not?
Just by calling `$this->any()`?

Comment (ihor-sviziev) [reply app/code/Magento/Catalog/Test/Unit/Model/Product/Option/SaveHandlerTest.php]: If you don’t need to check how many times it was called - you can use `any`, if you want to check that method was called - `atLeastOnce`, if you want to check if it was called only once or other amount of times - `once` or `exact`, but last case is needed only if this method is quite heavy, for instance doing load of model or runs DB query that running more times could lead to performance issues

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by oleksii-lisovyi** [comment]
Would be great to cover these changes with MFTF tests. I have no experience with them.
Could please somebody help with it?

**Comment by ihor-sviziev** [comment]
I think someone could get covering this change with MFTF using the following instruction https://github.com/magento/magento2/wiki/Test-coverage-contributions

**Comment by engcom-Kilo** [comment]
Working with MFTF test.

**Comment by oleksii-lisovyi** [comment]
@magento run all tests

**Comment by sidolov** [comment]
@magento create issue

**Comment by engcom-Kilo** [comment]
Hi @oleksii-lisovyi ,
please look on failed Static and Integration tests.
Thanks

**Comment by engcom-Kilo** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Kilo** [comment]
Added MFTF.

**Comment by oleksii-lisovyi** [comment]
@magento run Static Tests, Integration Tests

**Comment by engcom-Kilo** [comment]
@magento run Functional Tests B2B

**Comment by ihor-sviziev** [comment]
@magento run Integration Tests, Static Tests

**Comment by ihor-sviziev** [comment]
@magento run Integration Tests, Static Tests

**Comment by ihor-sviziev** [comment]
@magento run Database Compare, Functional Tests B2B, Functional Tests CE, Functional Tests EE, Magento Health Index, Sample Data Tests B2B, Sample Data Tests CE, Sample Data Tests EE, Semantic Version Checker, Unit Tests, WebAPI Tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8083](https://jira.corp.magento.com/browse/ENGCOM-8083) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Preconditions:**

Configure Magento in multi-store mode.
Create a website with 2 store views in it.

**Manual testing scenario:**

1. Create a product in Admin panel that supports custom options.
1.1. Assign a product to the website.
1.2. Add/import some custom options.
2. Save the product.
3. Change the config scope to store view level.
4. Change some of the attributes available on store view (e.g. price).
5. Save the product (don't open Custom Options tab).
6. Open Custom Options tab.

**Before:** :heavy_multiplication_x: Custom options are re-defined on store view level.

Screenshot from 2020-08-31 13-20-35

**After:** :heavy_check_mark: Custom options are not re-defined on store view level since nothing was actually changed.

Screenshot from 2020-08-31 13-26-14

**Comment by ihor-sviziev** [comment]
@engcom-Alfa 
Could you check additionally following cases?
1. please check if saving custom options still works fine via API
2. seems like this PR should also solve https://github.com/magento/magento2/issues/29572. Could you check it?

**Comment by engcom-Alfa** [comment]
Hi @ihor-sviziev . 

Was checked 2 additional cases:

1. check if saving custom options still works fine via API

:heavy_check_mark: GET, POST, DELETE, PUT actions with custom options via API were checked and everything works as expected. Custom options weren't re-defined on the store view level since nothing was actually changed.

2. and this PR also solve the issue #29572 

**Manual testing scenario:**

1. Add custom options to any product.
2. Update product SKU programmatically (custom script or custom module) that have custom options, for instance using following code:
```php
$productCollectionFactory = $objectManager->get('\Magento\Catalog\Model\ResourceModel\Product\CollectionFactory');
$storeId = 1;
$productCollection = $productCollectionFactory->create()->addAttributeToSelect('*')->addStoreFilter($storeId);
foreach ($productCollection as $product) {
    $sku = $product->getSku();
    $product->setSku('UpdatedSku-'.$sku);
    $product->save();
}
```
3. check the same product in the backend

**Actual Result:** :heavy_check_mark:  Product Sku was changed



**Actual Result:** :heavy_check_mark: The custom options aren't changed/deleted

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Review by engcom-Charlie — CHANGES_REQUESTED** [review]
@oleksii-lisovyi, `CreateOrderWithPaymentOnAccountTest` constantly falls on B2B.
Could you please take a look?
Thank you.

**Comment by ihor-sviziev** [comment]
@engcom-Charlie - @oleksii-lisovyi don't have access to B2B, could you help with failing test?

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
Risk set to medium since changes may affect critical features (product/reindex).

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8083](https://jira.corp.magento.com/browse/ENGCOM-8083) has been created to process this Pull Request

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

### Case 1

**Preconditions:**

Configure Magento in multi-store mode.
Create a website with 2 store views in it.

**Manual testing scenario:**

1. Create a product in Admin panel that supports custom options.
1.1. Assign a product to the website.
1.2. Add/import some custom options.
2. Save the product.
3. Change the config scope to store view level.
4. Change some of the attributes available on store view (e.g. price).
5. Save the product (don't open Custom Options tab).
6. Open Custom Options tab.

**After:** :heavy_check_mark: Custom options are not re-defined on store view level since nothing was actually changed.



### Case 2

:heavy_check_mark: GET, POST, DELETE, PUT actions with custom options via API were checked and everything works as expected. Custom options weren't re-defined on the store view level since nothing was actually changed.

### Case 3. and this PR also solve the issue #29572 

**Manual testing scenario:**

1. Add custom options to any product.
2. Update product SKU programmatically (custom script or custom module) that have custom options, for instance using following code:
```php
$productCollectionFactory = $objectManager->get('\Magento\Catalog\Model\ResourceModel\Product\CollectionFactory');
$storeId = 1;
$productCollection = $productCollectionFactory->create()->addAttributeToSelect('*')->addStoreFilter($storeId);
foreach ($productCollection as $product) {
    $sku = $product->getSku();
    $product->setSku('UpdatedSku-'.$sku);
    $product->save();
}
```
3. check the same product in the backend

**Actual Result:** :heavy_check_mark:  Product Sku was changed



**Actual Result:** :heavy_check_mark: The custom options aren't changed/deleted

**Review by engcom-Charlie — APPROVED** [review]

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B

**Comment by gabrieldagama** [comment]
@magento run all tests

**Review by gabrieldagama — CHANGES_REQUESTED** [review]
Hi @oleksii-lisovyi, thanks for contributing!

The B2B functional tests are constantly failing with this PR changes, can you have a look?

Thanks!

**Comment by ihor-sviziev** [comment]
@gabrieldagama AFAIK @oleksii-lisovyi don't have access to B2B repo.
@engcom-Charlie could you help with failing Functional B2B tests?

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@engcom-Charlie could you help with failing tests?

**Review by gabrieldagama — CHANGES_REQUESTED** [review]
@oleksii-lisovyi can you fix the conflicts? Thanks!

**Comment by ishakhsuvarov** [comment]
Hi @oleksii-lisovyi 

I am closing this PR for now as it seems like it has become stale and still needs work.

Please feel free to reopen if you wish to continue.

Thanks!

**Pull request closed without merging**